docs/transformations/ConstantFolding.md

```diff
@@ -0,0 +1,49 @@
+# Constant folding
+
+### Table of contents
+
+- [Overview](#overview)
+- [Working cases](#working-cases)
+- [Algorithm](#algorithm)
+
+### Overview
+
+This transformation aims to evaluate as much numerical/boolean constants
+as possible.
+
+### Working cases
+
+#### Fully constant integral/boolean expressions in any context
+
+For example:
+- `1 + 2 * 3 - 8 // 3 + 7 ^ 3 | 11` is replaced with `15`
+- `True and 1 + 1 == 2` is replaced with `True`
+
+*Note:* The power operator `**` is special in that the size of its result
+can be much larger than the size of its operands. Because of this,
+`**` is only evaluated when both its operands and its result fit into
+the JVM Long type (64-bits signed integer).
+
+#### Maximal (by inclusion) subexpressions of non-constant expressions
+
+For example:
+- `1 + 2 + x` is replaced with `3 + x`
+- `x + 1 + 2` is left unchanged (`+` is left-associative,
+  so there is no non-trivial constant subexpression)
```

Comment:
Is it true? We consider only `int` and `bool` variables and we can be sure that `x + 1 + 2 = x + 3`, is not?

Reply:
In this case, we would want to ensure type constraints on `x` and, right now, this kind of type information is not used at all... Also, we would be forced to use more complicated logic than simple recursive evaluation, since this expression is equivalent to `(x + 1) + 2`... I could try to implement this if we really need it :thinking:

Reply:
We should not know the type of `x` variable. I mean we can scan the subtree, reorder variables and calculate only `int` constants. For example,
`x + 1 + 3 + y + 8 + z - 5 - 1.5`
we can replace into `1 + 3 + 8 - 5 + x + y + z - 1.5 = 7 + x + y + z - 1.5` (or `x + y + z - 1.5 + 1 + 3 + 8 - 5 = x + y + z - 1.5 + 7` if you want
Here we keep the old order for everything except integer constants

Reply:
I don't think it's *really* equivalent without the type information :thinking: For example, `MyList` could be a custom list type where `lst + 3` means `lst.append(3)`. Then `MyList() + 1 + 2` would be `[1, 2]` and not `[3]`

Reply:
hm, I don't like Python 😢 PSI does not store type information in this case? I understand that there is a dynamic type system, but still

Reply:
Our сolleagues had the same problem. Let's use their approach to define a type in case we know what the type is and don't change the original expression otherwise
See, please [this](https://github.com/JetBrains-Research/python-ml-type-hints/blob/sketch/types_extractor/src/main/kotlin/extractor/FileTypesExtractor.kt#L97) link